memgraph-graphRAG/graphRAG.ipynb

```diff
@@ -0,0 +1,797 @@
+    "to execute in Memgraph."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_cypher_queries(nodes, relationships):\n",
+    "    queries = []\n",
+    "\n",
+    "    # Create nodes\n",
+    "    for node in nodes:\n",
+    "        query = f\"\"\"\n",
+    "        MERGE (n:{node['type']}:Entity {{name: '{node['name']}'}}) \n",
+    "        ON CREATE SET n.id={node['id']} \n",
+    "        ON MATCH SET n.id={node['id']}\n",
+    "        \"\"\"\n",
+    "        queries.append(query)\n",
+    "\n",
+    "    # Create relationships\n",
+    "    for rel in relationships:\n",
+    "        query = f\"MATCH (a {{id: {rel['source']}}}), (b {{id: {rel['target']}}}) \" \\\n",
+    "                f\"CREATE (a)-[:{rel['relationship']}]->(b)\"\n",
+    "        queries.append(query)\n",
+    "\n",
+    "    return queries\n",
+    "\n",
+    "cypher_queries = generate_cypher_queries(nodes, relationships)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Execute queries\n",
+    "\n",
+    "The final step is to execute those queries in Memgraph, enriching your graph\n",
+    "with the newly created context. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with driver.session() as session:\n",
+    "    for query in cypher_queries:\n",
+    "        try:\n",
+    "            session.run(query)\n",
+    "            print(f\"Executed query: {query}\")\n",
+    "        except Exception as e:\n",
+    "            print(f\"Error executing query: {query}. Error: {e}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataset now has additional knowledge: "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "vscode": {
+     "languageId": "sql"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
+    "MATCH (a {id: 1}), (b {id: 2}) CREATE (a)-[:PARENT_OF]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 3}) CREATE (a)-[:SIBLING_OF]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 4}) CREATE (a)-[:LOYAL_TO]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 5}) CREATE (a)-[:HAPPENED_IN]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 6}) CREATE (a)-[:HAPPENED_IN]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 7}) CREATE (a)-[:SIBLING_OF]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 8}) CREATE (a)-[:MARRIED_TO]->(b)\n",
+    "MATCH (a {id: 1}), (b {id: 9}) CREATE (a)-[:HEALED_BY]->(b)\n",
+    "MATCH (a {id: 8}), (b {id: 9}) CREATE (a)-[:RULES]->(b)\n",
+    "MATCH (a {id: 8}), (b {id: 1}) CREATE (a)-[:BETRAYED_BY]->(b)\n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "Now again asking the same question yields a correct answer: **\"Who betrayed Viserys Targaryen in Game of Thrones?\"**\n",
```

Comment:
There is a disconnect between this part and the last part in Ante's section. It is unclear what the wrong answer was and how this helped.